man/swup.1
.\" -*- nroff -*-
.\" $Id: swup.1,v 1.4 2001/03/13 14:01:07 olafb Exp $
.\" Copyright (c) Trustix AS 2001
.TH SWUP 1 "Jan 11, 2001"
.SH NAME
swup \- SoftWare UPdater
.SH SYNOPSIS
.B swup [operation] [options]
.SH DESCRIPTION
.B swup
is short for ``Software Updater''. It is an
extension to existing software package systems to facilitate automatic
install and upgrade.
.SH OPERATIONS
.IP "--help"
Print brief usage information.
.IP "--version"
Print version and copyright information.
.IP "--list-latest"
List latest version of packages available.
.IP "--list-new"
Like list-latest, but only list new packages that are not installed.
.IP "--install"
Install modus. Must be used together with a --package-* switch.
.IP "--upgrade"
Upgrade modus. Optional --package-* switch can be supplied.
.IP "--import-key"
Import a gpg public key. Takes one argument
.I filename
which is a file containing gpg public keys to be imported to keyring.
.IP "--list-keys"
List all public keys in gpg keyring.
.SH OPTIONS
.IP "--verbose"
Write somewhat more verbose messages to stdout.
.IP "--quiet"
Don't write status messages to stdout. Overrides --verbose.
.IP "--silent"
Be quiet and do not write any messages other than errors. Overrides
--verbose and --quiet.
.IP "--no-flush-cache"
Do not delete cached files after they are installed. Default is to
flush cache.
.IP "--poll-only"
Only poll for upgrades. This will print pending upgrades, but will not
try to resolve additional dependencies introduced by an upgrade. Used
in conjunction with --upgrade.
.IP "--download-only"
Run as usual, download packages, but do not install. Implies --no-flush-cache.
.IP "--save-to dir"
Save a copy of all downloaded software packages to dir.
.IP "--config-file file"
Use an alternate system wide
.I config-file
instead of
.I /etc/swup/swup.conf
File should not contain non-alphanumeric characters.
.IP "--repository-URI URI"
Use Uniform Resource Allocator to file that contains a list of latest
packages.  This will override the list of default URIs in
configuration file. Protocols
.I file,
.I http,
and
.I ftp
are supported.
.IP "--package-URI URI [URI]"
Use URIs to SPI for packages and try to install/upgrade these
specific packages.
.IP "--package-URI-file [...]"
Same as above, except that the URI-list is read from a file of newline
separated URIs.
.IP "--package name [...]"
Install/upgrade packages with name
.I name.
The name is shortname for the package. E.g. vim, and not
vim-5.6-3tr.i586.rpm.
.IP "--package-file file"
Same as above, except that the package names are given in a file,
separated by newline.
.IP "--gnupg-dir dir"
Use
.I dir
as directory where Gnu Privacy Guard files are stored, instead of the
default
.I /etc/swup/gnupg/
.SH FILES
.I /etc/swup/swup.conf
.RS
System wide configuration file for swup. See
.BR swup(5)
for more details.
.RE
.I /etc/swup/gnupg
.RS
Directory where Gnu Privacy Guard files are stored.
.SH DIAGNOSTICS
.I swup
will write messages to stdout and stderr. Upon completion a standard
exit code is returned to the calling shell. swup returns 0 for a
succesfull execution, 1 for general error and 2 for command line
syntax error.
.SH BUGS
None known.
.SH AUTHOR
Olaf Trygve Berglihn <olafb@trustix.com>




CHANGELOG
* 0.1.5
- Changed syntax to work with python 2.0

* 0.1.4
- Fixed bug that printed "this should never happen" when doing --list-keys, etc.

* 0.1.3
- Minor bugfixing.

* 0.1.2
- Fixed bug that ignored regexps for sites.

* 0.1.1
- Minor bugfixing.

* 0.1.0
- Changed config file format for sites. Configuration file parsing is not
  backwards compatible.
- General code cleanup and better exception handling.

* 0.0.8
- Added support for gz'ed files.
- Now checking md5sum on latest-files and cache parsed data.

* 0.0.7
- Added a new operation '--list-new' which lists available packages
  that are not already installed.
- Added a check for installed multiple versions of packages. Will now
  refuse to upgrade a package if multiple versions are installed. Will now
  also refuse to install if a package with same name is already installed.
- Added --save-to flag and save_to variable in config. If
  variable/switch is defined and is a name of an existing directory
  swup will try to save a copy of all rpms downloaded to the
  directory.


src/swup.py
#!/usr/bin/env python

import sys, string, os, re
import help, cmdline, config, log

true, false = 1, 0

# Exceptions
class SwupException(Exception):
    def __init__(self, value=None):
        self.value = value
    def __str__(self):
        return self.value

class ConfigException(SwupException): pass
class CmdlineException(SwupException): pass
class SetupException(SwupException): pass
class ImportKeyException(SwupException): pass
class ListKeysException(SwupException): pass


def parse_argv(argv):
    '''Parse command line arguments. Return dictionary with switch as key.'''
    try:
        argdict = cmdline.parse(argv)
    except cmdline.unknown, option:
        raise CmdlineException, "Unknown option: %s" % option
    except cmdline.wrong_number_arguments, switch:
        raise CmdlineException, "Wrong number of arguments for switch: %s"\
              % switch
    except cmdline.too_many_operations, (operation1, operation2):
        raise CmdlineException, "Too many operations: %s %s" %\
              (operation1, operation2)
    except cmdline.too_few_operations, message:
        raise CmdlineException, "Too few operations: %s" % message
    return argdict


def import_key(filename, gnupgdir):
    '''Given a local filename that contains a gnupg key and the dictionary
    in which the keyring is stored, import key into keyring.'''
    if os.path.isfile(filename):
        try:
            import gpg
            (retval, output) = gpg.import_key(filename, gnupgdir)
            sys.stdout.write("Import of key(s) succeded.\n")
        except gpg.key_import_failure, errmsg:
            raise ImportKeyException, errmsg
    else:
        raise ImportKeyException, "No such file '%s'" % filename


def list_keys(gnupgdir):
    '''Given a directory where the gnupg keyring is stored, list the keys
    in the keyring.'''
    try:
        import gpg
        (retval, output) = gpg.list_keys(gnupgdir)
        sys.stdout.write(output)
    except gpg.list_keys_failure, errmsg:
        raise ListKeysException, errmsg
    
    
def setup(argdict):
    '''Analyze argument dictionary, return (mode, args, config)
    where mode is a string, args is a list of strings and config is
    a dictionary of configuration variables with variable name as key.'''
	
    # did user request help or version info?
    if argdict.has_key('--help'):
        return ('help', None, None)
    if argdict.has_key('--version'):
        return ('version', None, None)

    # get config
    if argdict.has_key('--config-file'):
	config_file = argdict['--config-file'][0]
    else:
	config_file = config.DEFAULT_CONFIG_FILE

    configuration = None
    try:
	configuration = config.parse(config_file)
    except config.ConfigParseError, message:
        raise SetupException, '%s\n%s' % \
              ('Error in config file %s.' % config_file,
               'Parser said %s.' % message)

    # merge command line options with config
    if argdict.has_key('--silent'):
	configuration['loglevel'] = log.LOGLEVEL_SILENT 
    elif argdict.has_key('--quiet'):
        configuration['loglevel'] = log.LOGLEVEL_QUIET
    elif argdict.has_key('--verbose'):
        configuration['loglevel'] = log.LOGLEVEL_VERBOSE
    else:
        configuration['loglevel'] = log.LOGLEVEL_NORMAL
        
    if argdict.has_key('--no-flush-cache'):
	configuration['flush_cache'] = 'no'
    if argdict.has_key('--repository-URI'):
	configuration['sites'] = [('0', argdict['--repository-URI'][0],
				       '.*')]
    if argdict.has_key('--gnupg-dir'):
	configuration['gnupgdir'] = argdict['--gnupg-dir'][0]
    if argdict.has_key('--save-to'):
        todir = argdict['--save-to'][0]
        if not os.path.isdir( todir ):
            raise SetupException, "No such directory '%s'" % todir
        else:
            configuration['save_to'] = todir
        
    if argdict.has_key('--import-key'):
	filename = argdict['--import-key'][0]
        return ('import_key', filename, configuration)


    if argdict.has_key('--list-keys'):
        return ('list_keys', None, configuration)


    poll_only = false
    download_only = false

    if argdict.has_key('--poll-only'):
	configuration['poll_only'] = 'yes'
    if argdict.has_key('--download-only'):
	configuration['download_only'] = 'yes'
	configuration['flush_cache'] = 'no'

    packages = []

    if argdict.has_key('--package'):
	packages.extend(argdict['--package'])
    if argdict.has_key('--package-file'):
	try:
	    package_file = open(argdict['--package-file'][0])
	    package_file_list = package_file.readlines()
	    #package_file_list.remove('')
	    packages.extend(package_file_list)
	    package_file.close()
	except IOError: 
            raise SetupException, "Unable to open file: %s" %\
                  argdict['--package-file'][0]

    if argdict.has_key('--package-URI'):
	packages.extend(argdict['--package-URI'])

    if argdict.has_key('--package-URI-file'):
	try:
	    package_uri_file = \
			     open(argdict['--package-URI-file'][0])
	    package_uri_file_list = package_uri_file.readlines()
	    #package_uri_file_list.remove("")
	    packages.extend(package_uri_file_list)
	    package_uri_file.close()
	except IOError: 
            raise SetupException, "Unable to open URI-file: %s" %\
                  argdict['--package-file'][0]

    # prettify list
    for i in range(len(packages)):
	packages[i] = string.strip(packages[i])

    # create the installation object

    if argdict.has_key('--upgrade'):
	if packages == []:
            mode = 'upgrade_normal'
	else:
            mode = 'upgrade_package'
        return (mode, packages, configuration)

    elif argdict.has_key('--install'):
	if packages == []:
            raise SetupException, \
                  'You need to specify at least one package to install.'
        return ('install', packages, configuration)
    
    elif argdict.has_key('--list-latest'):
        return ('list_latest', None, configuration)

    elif argdict.has_key('--list-new'):
        return ('list_new', None, configuration)

    else:
        raise SetupExeption, "This should not happen."


def main(argv):
    'Main interface. Takes list of command line arguments as argument.'

    # Parse the command line arguments.
    try:
        argdict = parse_argv(argv)
    except CmdlineException, errmsg:
        sys.stderr.write('%s\n' % errmsg)
        sys.exit(2)

    # Setup the configuration and get the mode in which we are supposed to
    # run.
    try:
        mode, args, config = setup(argdict)
    except SetupException, errmsg:
        sys.stderr.write('%s\n' % errmsg)
        sys.exit(2)

    # Check for the simpler modes that does not involve the upgrade module.
    if mode == 'help':
        help.helpme()
        sys.exit(0)
    elif mode == 'version':
        help.version()
        sys.exit(0)
    elif mode == 'import_key':
        try:
            import_key(args, config['gnupgdir'])
            sys.exit(0)
        except ImportKeyException, errmsg:
            sys.stderr.write('%s\n' % errmsg)
            sys.exit(2)
    elif mode == 'list_keys':
        try:
            list_keys(config['gnupgdir'])
            sys.exit(0)
        except ListKeysException, errmsg:
            sys.stderr.write('%s\n' % errmsg)
            sys.exit(1)

    # The mode requires the upgrade module. Initialize an upgrade agent
    # and run.

    import upgrade
    agent = None
    try:
        if mode == 'upgrade_normal':
            agent = upgrade.Upgrade(config)
        elif mode == 'upgrade_package':
            agent = upgrade.upgrade_package(args, config)
        elif mode == 'install':
            agent = upgrade.install_package(args, config)
        elif mode == 'list_latest':
            agent = upgrade.list_latest(config)
        elif mode == 'list_new':
            agent = upgrade.list_new(config)
    except upgrade.UpgradeException, err:
        sys.stderr.write('Failed to initialize: %s\n' % err)
        sys.exit(1)
        
    try:
        if agent:
            agent.run()
        else:
            sys.stderr.write('This should never happen.\n')
    except upgrade.UpgradeException:
        sys.exit(1)

    # All went ok if we got here. Exit normally.
    sys.exit(0)


if __name__ == '__main__':
    main(sys.argv[1:])







